### PR TITLE
batch apply

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
This pr will reduce CPU usage for apply worker, I will introduce sysbench result later.

### CODE REVIEW

File: src/raftstore/store/worker/apply.rs [hunk 1]
```diff
@@ -441,28 +461,33 @@ impl ApplyDelegate {
...
     }
 }
 
-struct ExecContext<'a> {
+struct ExecContext<'a, 'b> {
     snap: Snapshot,
     apply_state: RaftApplyState,
-    wb: WriteBatch,
-    req: &'a RaftCmdRequest,
+    wb: &'a mut WriteBatch,
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: seem WriteBatch and request can have same lifetime.

File: src/raftstore/store/worker/apply.rs [hunk 2]
```diff
@@ -661,9 +686,9 @@ impl ApplyDelegate {
         let region_ver = region.get_region_epoch().get_version() + 1;
         region.mut_region_epoch().set_version(region_ver);
         new_region.mut_region_epoch().set_version(region_ver);
-        write_peer_state(&ctx.wb, &region, PeerState::Normal)
-            .and_then(|_| write_peer_state(&ctx.wb, &new_region, PeerState::Normal))
-            .and_then(|_| write_initial_state(self.engine.as_ref(), &ctx.wb, new_region.get_id()))
+        write_peer_state(ctx.wb, &region, PeerState::Normal)
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: we can save only the peer state once in the end.

**Comment by zhangjinpeng1987** [comment]
sysbench insert result:
1 pd 1 tidb 3 tikv 
256 threads, 16 tables, each table 1000,000 rows
```
branch 
   transactions:                        3755479 (6258.80 per sec.)
   queries:                             3755479 (6258.80 per sec.)
   min:                                  3.02
   avg:                                 40.90
   max:                               1632.18
   95th percentile:                     71.83
master
    transactions:                        3354794 (5590.65 per sec.)
    queries:                             3354794 (5590.65 per sec.)
    min:                                  3.02
    avg:                                 45.79
    max:                                487.48
    95th percentile:                     87.56
```
apply worker CPU usage reduce by about 5% in sysbench insert case.

**Comment by siddontang** [comment]
The transaction QPS seems to improve 15%, and the latency seems to reduce too, awesomely.

File: src/raftstore/store/worker/apply.rs [hunk 3]
```diff
@@ -132,8 +132,12 @@ fn notify_stale_command(tag: &str, term: u64, mut cmd: PendingCmd) {
...
 }
 
+fn gen_stale_req(uuid: Uuid, term: u64) -> RaftCmdResponse {
+    cmd_resp::err_resp(Error::StaleCommand, uuid, term)
+}
+
 pub fn notify_stale_req(tag: &str, term: u64, uuid: Uuid, cb: Callback) {
-    let resp = cmd_resp::err_resp(Error::StaleCommand, uuid, term);
+    let resp = gen_stale_req(uuid, term);
     info!("{} command {} is stale, skip", tag, uuid);
```
Comment (BusyJay) [inline src/raftstore/store/worker/apply.rs]: The log can be moved to L137.

File: src/raftstore/store/worker/apply.rs [hunk 4]
```diff
@@ -209,64 +216,87 @@ impl ApplyDelegate {
...
+
+        self.metrics.written_bytes += wb.data_size() as u64;
+        self.metrics.written_keys += wb.count() as u64;
+
+        // Write to storage and call callbacks
+        self.engine
+            .write(wb)
+            .unwrap_or_else(|e| panic!("{} failed to write to engine, error: {:?}", self.tag, e));
+        for (cb, resp) in cbs {
+            cb.call_box((resp,));
```
Comment (BusyJay) [inline src/raftstore/store/worker/apply.rs]: prefer cb(resp).

File: src/raftstore/store/worker/apply.rs [hunk 5]
```diff
@@ -209,64 +216,87 @@ impl ApplyDelegate {
...
+                }
             };
 
             if let Some(res) = res {
                 results.push(res);
             }
         }
 
+        if !self.pending_remove {
+            rocksdb::get_cf_handle(&self.engine, CF_RAFT)
```
Comment (hhkbp2) [inline src/raftstore/store/worker/apply.rs]: It's better to add a function or add a comment for this block.

Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: L199 has comment for `pending_remove`.
I think it is clearly that if this peer is remove itself, no need write status to rocksdb.

**Comment by hhkbp2** [comment]
Rest LGTM

To guarantee service latency in worse case, there should be a limit on the write batch size. To get the best tradeoff between QPS and average/max latency, it needs to tweak the batch size. There is still some space for improvement in the future.

**Comment by siddontang** [comment]
PTAL @BusyJay

**Comment by BusyJay** [comment]
LGTM

**Comment by siddontang** [comment]
CI failed @zhangjinpeng1987

**Comment by zhangjinpeng1987** [comment]
PTAL

File: src/raftstore/store/worker/apply.rs [hunk 6]
```diff
@@ -209,64 +216,112 @@ impl ApplyDelegate {
...
-                EntryType::EntryNormal => self.handle_raft_entry_normal(entry),
-                EntryType::EntryConfChange => self.handle_raft_entry_conf_change(entry),
+                EntryType::EntryNormal => self.handle_raft_entry_normal(&mut wb, &mut cbs, entry),
+                EntryType::EntryConfChange => {
+                    self.handle_raft_entry_conf_change(&mut wb, &mut cbs, entry)
+                }
             };
 
+            // When encounter ComputeHash cmd, we must flush the writebatch to engine immediately.
+            if let Some(ref exec_result) = res {
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: should extract a function to check whether flush forcibly or not.

**Comment by siddontang** [comment]
Rest LGTM

But CI still failed @zhangjinpeng1987

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/raftstore/store/worker/apply.rs [hunk 7]
```diff
@@ -132,12 +132,26 @@ fn notify_stale_command(tag: &str, term: u64, mut cmd: PendingCmd) {
...
     info!("{} command {} is stale, skip", tag, uuid);
+    cmd_resp::err_resp(Error::StaleCommand, uuid, term)
+}
+
+pub fn notify_stale_req(tag: &str, term: u64, uuid: Uuid, cb: Callback) {
+    let resp = gen_stale_req(tag, uuid, term);
     cb(resp);
 }
 
+fn should_flush_to_engine_immediately(res: &Option<ExecResult>) -> bool {
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: em, I think immediately is not proper in the function name, can be removed.

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL again

File: src/raftstore/store/worker/apply.rs [hunk 8]
```diff
@@ -209,64 +226,108 @@ impl ApplyDelegate {
...
             let res = match entry.get_entry_type() {
-                EntryType::EntryNormal => self.handle_raft_entry_normal(entry),
-                EntryType::EntryConfChange => self.handle_raft_entry_conf_change(entry),
+                EntryType::EntryNormal => self.handle_raft_entry_normal(&mut wb, &mut cbs, entry),
+                EntryType::EntryConfChange => {
+                    self.handle_raft_entry_conf_change(&mut wb, &mut cbs, entry)
+                }
             };
 
+            if should_flush_to_engine(&res) {
```
Comment (BusyJay) [inline src/raftstore/store/worker/apply.rs]: Snapshot is still stale.

Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: how to fix it?

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL
I aggr multi regions to a write batch, this can reduce many cpu when there are many active regions.

**Comment by siddontang** [comment]
@BusyJay  @hhkbp2 

Please review it again.

File: src/raftstore/store/worker/apply.rs [hunk 9]
```diff
@@ -39,6 +39,7 @@ use raftstore::store::peer_storage::{self, write_initial_state, write_peer_state
 use raftstore::store::peer::{parse_data_at, check_epoch, Peer};
 use raftstore::store::metrics::*;
 
+const WRITE_BATCH_RECOMMENDED_KEYS: usize = 128;
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: WRITE_BATCH_MAX_KEYS

Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: 128 is small?

File: src/raftstore/store/worker/apply.rs [hunk 10]
```diff
@@ -208,78 +254,114 @@ impl ApplyDelegate {
...
+
         slow_log!(t,
                   "{} handle ready {} committed entries",
                   self.tag,
                   committed_count);
         results
     }
 
-    fn handle_raft_entry_normal(&mut self, entry: Entry) -> Option<ExecResult> {
+    fn write_apply_state(&mut self, wb: &mut WriteBatch) {
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: dodo we need `mut` here?

File: src/raftstore/store/worker/apply.rs (hunk 9 above)
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: As we have benched, when the write batch grow over about a hundred keys, the extra benefits is very trivial, on the other hand we should response the caller of raftkv asap, so we use 128 as the max keys one write batch should contains.

File: src/raftstore/store/worker/apply.rs [hunk 11]
```diff
@@ -208,78 +254,114 @@ impl ApplyDelegate {
...
+
         slow_log!(t,
                   "{} handle ready {} committed entries",
                   self.tag,
                   committed_count);
         results
     }
 
-    fn handle_raft_entry_normal(&mut self, entry: Entry) -> Option<ExecResult> {
+    fn write_apply_state(&self, wb: &mut WriteBatch) {
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: seem no need mut for WriteBatch here.

Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: Mut indicate that will write something to writebatch.

Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: but the definition for the Write API of WriteBatch is not `mut`, so we don't need it at all.

Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: The definition for the write abi of WriteBatch is not unreasonable, I will change it in another pr.

Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: This may be reasonable or unreasonable, many other libraries also use this way for the FFI API, you can even see some Rust API uses immutable self too, like TcpStream `SetXXX`. So I think this is ok here, and we should follow the style and keep consistent now.

File: src/raftstore/store/worker/apply.rs [hunk 12]
```diff
@@ -110,6 +111,28 @@ pub enum ExecResult {
...
+
+impl ApplyContext {
+    fn new() -> ApplyContext {
+        ApplyContext {
+            wb: Some(WriteBatch::new()),
+            cbs: vec![],
+        }
+    }
+
+    pub fn wb_mut(&mut self) -> &mut WriteBatch {
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: can remove this function now.

File: src/raftstore/store/worker/apply.rs [hunk 13]
```diff
@@ -208,78 +254,114 @@ impl ApplyDelegate {
...
+                    .write(apply_ctx.wb.take().unwrap())
+                    .unwrap_or_else(|e| {
+                        panic!("{} failed to write to engine, error: {:?}", self.tag, e)
+                    });
+
+                // call callback
+                for (cb, resp) in apply_ctx.cbs.drain(..) {
+                    cb(resp);
+                }
+                apply_ctx.wb = Some(WriteBatch::new());
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: seem we can use WriteBatch clear to reuse the allocated buffer.

Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: engine.write will take wb away.

File: src/raftstore/store/worker/apply.rs (hunk 12 above)
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: There are a lot of places use `&mut WriteBatch` to indicate that will modify wb, should I follow the style?
https://github.com/pingcap/tikv/blob/master/src/raftstore/store/peer_storage.rs#L180
https://github.com/pingcap/tikv/blob/master/src/raftstore/store/peer_storage.rs#L171
https://github.com/pingcap/tikv/blob/master/src/raftstore/store/peer_storage.rs#L530
https://github.com/pingcap/tikv/blob/master/src/raftstore/store/peer_storage.rs#L565

File: tests/raftstore/test_stats.rs [hunk 14]
```diff
@@ -137,9 +137,11 @@ fn must_detect_snap(pd_client: &Arc<TestPdClient>) {
     for _ in 0..200 {
         sleep_ms(10);
 
-        if let Some(stats) = pd_client.get_store_stats(1) {
-            if stats.get_sending_snap_count() > 0 || stats.get_receiving_snap_count() > 0 {
-                return;
+        for id in 1..3 {
+            if let Some(stats) = pd_client.get_store_stats(id) {
```
Comment (siddontang) [inline tests/raftstore/test_stats.rs]: do we need to use another PR for this?

**Comment by siddontang** [comment]
any test to cover `should_flush_to_engine`?

File: src/raftstore/store/worker/apply.rs (hunk 12 above)
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: oh sorry, I got wrong. You can follow your style.

**Comment by siddontang** [comment]
I suggest adding some tests to check the apply state. Maybe we can merge #1783 first and then add tests base on it.

**Comment by siddontang** [comment]
@zhangjinpeng1987 please fix the conflict and add some tests

**Comment by siddontang** [comment]
Any update? @zhangjinpeng1987

**Comment by siddontang** [comment]
Now PEER_APPLY_LOG_HISTOGRAM can't work as before, we should move it to the proper place.

File: src/raftstore/store/worker/apply.rs [hunk 15]
```diff
@@ -233,78 +303,122 @@ impl ApplyDelegate {
...
             if let Some(res) = res {
                 results.push(res);
             }
         }
 
+        if !self.pending_remove {
+            self.write_apply_state(apply_ctx.wb_mut());
+        }
+
+        self.update_metrics(apply_ctx);
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: why not update the keys and count when writing the DB with WriteBatch directly? 

Seem no need to use the delta mechanism here.

**Comment by siddontang** [comment]
PTAL @hhkbp2

File: src/raftstore/store/worker/apply.rs (hunk 15 above)
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: Because we merge multi region's write to one writebatch.

**Comment by zhangjinpeng1987** [comment]
@hhkbp2 PTAL

**Comment by hhkbp2** [comment]
LGTM

File: src/raftstore/store/worker/apply.rs [hunk 16]
```diff
@@ -233,78 +303,122 @@ impl ApplyDelegate {
...
         let index = entry.get_index();
         let term = entry.get_term();
         let data = entry.get_data();
 
         if !data.is_empty() {
             let cmd = parse_data_at(data, index, &self.tag);
-            return self.process_raft_cmd(index, term, cmd);
+
+            if should_flush_to_engine(&cmd, apply_ctx.wb_ref().count()) {
+                self.write_apply_state(apply_ctx.wb_mut());
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: do we have any case to cover this, not only for should_flush_to_engine?

Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: ???

Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: I have add test for this case.

**Comment by zhangjinpeng1987** [comment]
PTAL

File: src/raftstore/store/worker/apply.rs [hunk 17]
```diff
@@ -1106,40 +1210,62 @@ impl Runner {
...
-            }
-            MapEntry::Occupied(e) => e,
-        };
-        {
-            let delegate = e.get_mut();
-            delegate.metrics = ApplyMetrics::default();
-            delegate.term = apply.term;
-            let results = delegate.handle_raft_committed_entries(apply.entries);
+    fn handle_apply(&mut self, applys: Vec<Apply>) {
+        let timer = STORE_APPLY_LOG_HISTOGRAM.start_timer();
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: you can use `let _timer`, so you don't need to call `observe_duration` explicitly in the end of this function.

File: src/raftstore/store/worker/apply.rs [hunk 18]
```diff
@@ -1106,40 +1210,62 @@ impl Runner {
...
+                applys_res.push(ApplyRes {
                     region_id: apply.region_id,
                     apply_state: delegate.apply_state.clone(),
                     exec_res: results,
                     metrics: delegate.metrics.clone(),
                     applied_index_term: delegate.applied_index_term,
-                }))
-                .unwrap();
+                });
+
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: remove the blank lines

**Comment by siddontang** [comment]
PTAL @javaforfun

File: src/raftstore/store/worker/apply.rs (hunk 17 above)
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: so cool

**Comment by javaforfun** [comment]
LGTM

File: src/raftstore/store/worker/apply.rs [hunk 19]
```diff
@@ -1197,7 +1319,7 @@ impl Runner {
 impl Runnable<Task> for Runner {
     fn run(&mut self, task: Task) {
         match task {
-            Task::Apply(a) => self.handle_apply(a),
+            Task::Applys(a) => self.handle_apply(a),
```
Comment (BusyJay) [inline src/raftstore/store/worker/apply.rs]: `Applys` or `Applies`?
